neurom/apps/morph_stats.py

```diff
@@ -112,7 +122,13 @@ def extract_dataframe(neurons, config):
     if 'neuron' in config:
         del config['neuron']
 
-    stats = {nrn.name: extract_stats(nrn, config) for nrn in neurons}
+    func = partial(_run_extract_stats, config=config)
+    if n_workers == 1:
+        stats = dict(map(func, neurons))
+    else:
+        with multiprocessing.Pool(n_workers) as pool:
+            stats = dict(pool.imap_unordered(func, neurons))
```

Comment:
you never use the key ordering from the stats to do anything ?

Reply:
see line 132, it is converted to 'name' in line 134

Reply:
so in the final dataframe we have the name of the cells for which the features have been extracted

Reply:
Then I would have sorted the 
```
pd.DataFrame(columns=['name', 'neurite_type'] + columns, data=rows).sort(axis=1)
```

Reply:
this avoids a tonne of problems when you will try to compare stuffs

Reply:
sure, or just use .imap()

Reply:
> sure, or just use .imap() 
hum I don t think so. iirc, imap is blocking.

Reply:
yes, we don't care about blocking here

Reply:
ok up to you then

Reply:
I let you check if you really have the same results also with the single worker for the same input. 
this deserve a test.

Reply:
but we don't care about ordering with pandas, right? I avoid using .iloc unless really necessary.

Reply:
This is not for querying but for the comparisons :
```
>> import pandas as pd 
>> a = {"a": [1,2], "b" : [1,2]}   
>> b = {"b": [1,2], "a" : [1,2]}   
>> pd.DataFrame(a) == pd.DataFrame(b)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-8-311c305e4e29> in <module>
----> 1 pd.DataFrame(a) == pd.DataFrame(b)

~/.virtualenvs/snap/lib/python3.6/site-packages/pandas/core/ops/__init__.py in f(self, other)
    837             if not self._indexed_same(other):
    838                 raise ValueError(
--> 839                     "Can only compare identically-labeled DataFrame objects"
    840                 )
    841             new_data = dispatch_to_series(self, other, op, str_rep)

ValueError: Can only compare identically-labeled DataFrame objects
```
```
import pandas.testing as npt
npt.assert_frame_equal(pd.DataFrame(a), pd.DataFrame(b))                                                                                                                                           
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-12-cff663a5701a> in <module>
----> 1 npt.assert_frame_equal(pd.DataFrame(a), pd.DataFrame(b))

~/.virtualenvs/snap/lib/python3.6/site-packages/pandas/_testing.py in assert_frame_equal(left, right, check_dtype, check_index_type, check_column_type, check_frame_type, check_less_precise, check_names, by_blocks, check_exact, check_datetimelike_compat, check_categorical, check_like, obj)
   1350         check_exact=check_exact,
   1351         check_categorical=check_categorical,
-> 1352         obj=f"{obj}.columns",
   1353     )
   1354 

~/.virtualenvs/snap/lib/python3.6/site-packages/pandas/_testing.py in assert_index_equal(left, right, exact, check_names, check_less_precise, check_exact, check_categorical, obj)
    684             obj=obj,
    685             lobj=left,
--> 686             robj=right,
    687         )
    688 

pandas/_libs/testing.pyx in pandas._libs.testing.assert_almost_equal()

pandas/_libs/testing.pyx in pandas._libs.testing.assert_almost_equal()

~/.virtualenvs/snap/lib/python3.6/site-packages/pandas/_testing.py in raise_assert_detail(obj, message, left, right, diff)
    913         msg += f"\n[diff]: {diff}"
    914 
--> 915     raise AssertionError(msg)
    916 
    917 

AssertionError: DataFrame.columns are different

```

Reply:
mmh here you compare on columns ordering, not index, right? The sorting you want to do on 'name' is on indices of dataframe. But yes, it's cleaner to ensure the format of df is the same, we should add a .sort() at the end. (I never had to compare dataframes, but maybe one day, who knows!)

Reply:
You compare on both.

So the checks are : 
a.columns == b.columns which includes the ordering
a.index == b.index which includes the ordering
a.data == a.data

why ordering is important ? 
- because if you have (one day) multi level indexers (https://pandas.pydata.org/pandas-docs/stable/user_guide/advanced.html) the algorithm (and so the performances) are different with sorted or unsorted index.
- if you want to use `iloc` the results would differs


Be careful, you need to sort on both axis ! `.sort()` only sorts on indices and not columns.

Reply:
I did not think order was important that's why I transformed `imap` to `imap_unordered`. To me, `.sort()` only solves half of the issue.  Adding a `.sort()` would ensure stability, but it would not even ensure the result is the same as the one with running the code with `n_workers==1`. And it would not even be the order of the `neurons` list.
It think it's either:
- `imap_unordered` if we don't care about the ordering. I would say, we don't. morph_stat is about aggregating data, it will most of the time be grouped by mtype afterward, so the order will disappear anyway.
- switch back to `imap` if we do. 
- add a `sort` flag to switch between `imap` and `imap_unordered`

Reply:
+1 for leave it as it is

Reply:
>  Adding a .sort() would ensure stability, but it would not even ensure the result is the same as the one with running the code with n_workers==1. 

Should be or I am missing something ... And if the ordering makes sense is something else. 

But up to you again. I don t know the final goal of this code so maybe this is just pure pain for nothing indeed. I just share my experience with dataframes.